tree/ntuple/v7/src/RNTupleSerialize.cxx

```diff
@@ -739,23 +801,29 @@ std::uint32_t ROOT::Experimental::Internal::RNTupleSerializer::SerializeLocator(
    const RNTupleLocator &locator, void *buffer)
 {
    std::uint32_t size = 0;
-   if (!locator.fUrl.empty()) {
-      if (locator.fUrl.length() >= (1 << 24))
+   if (locator.IsSimple()) {
+      if (static_cast<std::int32_t>(locator.fBytesOnStorage) < 0)
          throw RException(R__FAIL("locator too large"));
-      std::int32_t head = locator.fUrl.length();
-      head |= 0x02 << 24;
-      head = -head;
-      size += SerializeInt32(head, buffer);
-      if (buffer)
-         memcpy(reinterpret_cast<unsigned char *>(buffer) + size, locator.fUrl.data(), locator.fUrl.length());
-      size += locator.fUrl.length();
+      size += SerializeUInt32(locator.fBytesOnStorage, buffer);
+      size += SerializeUInt64(locator.Get<std::uint64_t>(),
+                              buffer ? reinterpret_cast<unsigned char *>(buffer) + size : nullptr);
       return size;
    }
 
-   if (static_cast<std::int32_t>(locator.fBytesOnStorage) < 0)
-      throw RException(R__FAIL("locator too large"));
-   size += SerializeUInt32(locator.fBytesOnStorage, buffer);
-   size += SerializeUInt64(locator.fPosition, buffer ? reinterpret_cast<unsigned char *>(buffer) + size : nullptr);
+   const int type = std::holds_alternative<std::string>(locator.fPosition)
+                       ? static_cast<std::uint8_t>(ROOT::Experimental::RNTupleLocator::kTypeURI)
+                       : locator.fType;
+   if (!type)
+      throw RException(R__FAIL("locator has no type"));
+   auto payloadp = buffer ? reinterpret_cast<unsigned char *>(buffer) + sizeof(std::int32_t) : nullptr;
+   size += std::visit(
+      [&](auto &&value) { return SerializeLocatorPayload<std::decay_t<decltype(value)>>(locator, payloadp); },
```

Comment:
I would find it much more readable to simply use a switch statement on locator `fType`. And then to explicitly call `SerializeLocatorPayloadURI` or `SerializeLocatorPayloadObject64`

Reply:
In principle, all major compilers are able to optimize this out to a `switch` over the `index()` of the `std::optional`, with the benefit that if a new alternative is added, one gets a compile-time error if serialization for that type is missing:  
```c++
/home/jalopezg/CERN/repos/root/tree/ntuple/v7/src/RNTupleSerialize.cxx:310:49: error: ‘Serialize’ is not a member of ‘{anonymous}::LocatorPayloadSerializer<ROOT::Experimental::Foo>’
  310 |    return LocatorPayloadSerializer<T>::Serialize(std::forward<ArgsT>(args)...);
      |           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

However, I agree that _that_ makes the code more complicated.  I would agree to change it to a `switch` over `fType`, if `-Wswitch` is enabled so that we get at least a warning if there is not a `case` statement for all of the enumerators in `ELocatorType`.